hbase-backup/src/test/java/org/apache/hadoop/hbase/backup/TestIncrementalBackup.java

```diff
@@ -100,142 +141,227 @@ public void TestIncBackupRestore() throws Exception {
-      String backupIdIncMultiple2 = client.backupTables(request);
-      assertTrue(checkSucceeded(backupIdIncMultiple2));
-      validateRootPathCanBeOverridden(BACKUP_ROOT_DIR, backupIdIncMultiple2);
-
-      // #5 - restore full backup for all tables
-      TableName[] tablesRestoreFull = new TableName[] { table1, table2 };
-      TableName[] tablesMapFull = new TableName[] { table1_restore, table2_restore };
-
-      LOG.debug("Restoring full " + backupIdFull);
-      client.restore(BackupUtils.createRestoreRequest(BACKUP_ROOT_DIR, backupIdFull, false,
-        tablesRestoreFull, tablesMapFull, true));
-
-      // #6.1 - check tables for full restore
-      Admin hAdmin = TEST_UTIL.getAdmin();
-      assertTrue(hAdmin.tableExists(table1_restore));
-      assertTrue(hAdmin.tableExists(table2_restore));
-      hAdmin.close();
-
-      // #6.2 - checking row count of tables for full restore
-      Table hTable = conn.getTable(table1_restore);
-      Assert.assertEquals(HBaseTestingUtil.countRows(hTable), NB_ROWS_IN_BATCH + NB_ROWS_FAM3);
-      hTable.close();
-
-      hTable = conn.getTable(table2_restore);
-      Assert.assertEquals(NB_ROWS_IN_BATCH, HBaseTestingUtil.countRows(hTable));
-      hTable.close();
-
-      // #7 - restore incremental backup for multiple tables, with overwrite
-      TableName[] tablesRestoreIncMultiple = new TableName[] { table1, table2 };
-      TableName[] tablesMapIncMultiple = new TableName[] { table1_restore, table2_restore };
-      client.restore(BackupUtils.createRestoreRequest(BACKUP_ROOT_DIR, backupIdIncMultiple2, false,
-        tablesRestoreIncMultiple, tablesMapIncMultiple, true));
-      hTable = conn.getTable(table1_restore);
-
-      LOG.debug("After incremental restore: " + hTable.getDescriptor());
-      int countFamName = TEST_UTIL.countRows(hTable, famName);
-      LOG.debug("f1 has " + countFamName + " rows");
-      Assert.assertEquals(countFamName, NB_ROWS_IN_BATCH + ADD_ROWS);
-
-      int countFam2Name = TEST_UTIL.countRows(hTable, fam2Name);
-      LOG.debug("f2 has " + countFam2Name + " rows");
-      Assert.assertEquals(countFam2Name, NB_ROWS_FAM2);
-
-      int countMobName = TEST_UTIL.countRows(hTable, mobName);
-      LOG.debug("mob has " + countMobName + " rows");
-      Assert.assertEquals(countMobName, NB_ROWS_MOB);
-      hTable.close();
-
-      hTable = conn.getTable(table2_restore);
-      Assert.assertEquals(NB_ROWS_IN_BATCH + 5, HBaseTestingUtil.countRows(hTable));
-      hTable.close();
+      incrementalBackupId = backupAdmin.backupTables(request);
+      assertTrue(checkSucceeded(incrementalBackupId));
+
+      backupAdmin.restore(BackupUtils.createRestoreRequest(BACKUP_ROOT_DIR, incrementalBackupId,
+        false, fromTables, toTables, true, true));
+
+      table = conn.getTable(table1);
+      table.close();
+      int rowsExpected = HBaseTestingUtil.countRows(table, famName, mobFam);
+      table = conn.getTable(table1_restore);
```

Comment:
please use try-with-resource

Reply:
does try with resources make the code cleaner in this case? I might argue that it makes it messier

Reply:
either way, I think I can resolve these by applying Dieter's suggestion

Reply:
It may not be any more legible, indeed it may be worse. But this is about closing the resources in a finally block under failure scenario. We want the resources to be cleaned up as best we can so that subsequent tests run in this JVM have the best chance of succeeding. Our test suite is flakey enough without leaving orphaned resources about.